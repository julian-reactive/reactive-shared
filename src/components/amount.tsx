// Libraries
import React, { useState, useCallback } from 'react'

// Material components
import Box from '@mui/material/Box'
import Fab from '@mui/material/Fab'
import TextField from '@mui/material/TextField'

// Icons
import AddIcon from '@mui/icons-material/Add'
import RemoveIcon from '@mui/icons-material/Remove'

// Shared
import { onlyText } from '../utils'

interface AmountComponentProps {
  onPlus: (value: number) => void
  onMinus: (value: number) => void
  maxValue?: number
  minValue?: number
  initialValue?: number
}

const AmountComponent: React.FC<AmountComponentProps> = ({ onPlus, onMinus, maxValue = Number.MAX_SAFE_INTEGER, minValue = 0, initialValue = 0 }) => {
  const [amount, setAmount] = useState(initialValue)

  const handlePlus = useCallback(() => {
    setAmount(value => {
      if (value < maxValue) {
        value = value + 1
        onPlus(value)
      }
      return value
    })
  }, [onPlus, maxValue])

  const handleMinus = useCallback(() => {
    setAmount(value => {
      if (amount > minValue) {
        value = value - 1
        onMinus(value)
      }
      return value
    })
  }, [onMinus, amount, minValue])

  return (
    <Box sx={{ display: 'flex', justifyContent: 'center', py: 1 }}>
      <Fab
        aria-label='remove 1 unit amount'
        color='secondary'
        onClick={handleMinus}
        size='small'
      >
        <RemoveIcon fontSize='small' />
      </Fab>

      <TextField
        sx={{ px: 1, width: '40%' }}
        label={onlyText('FORM.LABEL.AMOUNT')}
        variant="outlined"
        value={amount}
        disabled
        size='small'
      />

      <Fab
        aria-label='add 1 unit amount'
        color='primary'
        onClick={handlePlus}
        size='small'
      >
        <AddIcon fontSize='small' />
      </Fab>
    </Box>
  )
}

export const SharedAmount = React.memo(AmountComponent)
