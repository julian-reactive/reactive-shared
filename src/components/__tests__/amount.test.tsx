import React from 'react'

import { render, screen } from '@testing-library/react'
import { SharedAmount, AmountComponentProps } from '../amount'
import { act } from 'react-dom/test-utils'
import { userEvent } from '@storybook/testing-library'

jest.mock('../../utils', () => ({
    onlyText: () => 'amount'
}))

const defaultProps: AmountComponentProps = {
  onPlus: () => 1,
  onMinus: () => -1,
  maxValue: 10,
  minValue: 0,
  initialValue: 1
}

const getRendered = (props = defaultProps) => render(<SharedAmount {...props}/>)

describe('amount', () => {
  it('should match snapshot', () => {
    const { container } = getRendered()
    expect(container.firstChild).toMatchSnapshot()
  })

  describe('when onPlus is clicked', () => {
    it('should add 1 to value', async () => {
      getRendered()

      const expectedValue = defaultProps.initialValue! + 1

      const textField = screen.getByLabelText('amount') as HTMLInputElement
      const plusButton = screen.getByLabelText('add 1 unit amount')

      expect(textField.value).toEqual(defaultProps.initialValue?.toString())

      act( () => {
         userEvent.click(plusButton)
      })

      expect(textField.value).toEqual(expectedValue.toString())
    })
  })

  describe('when onMinus is clicked', () => {
    it('should remove 1 to value', async () => {
      getRendered()

      const expectedValue = defaultProps.initialValue! - 1

      const textField = screen.getByLabelText('amount') as HTMLInputElement
      const minusButton = screen.getByLabelText('remove 1 unit amount')

      expect(textField.value).toEqual(defaultProps.initialValue?.toString())

      act( () => {
         userEvent.click(minusButton)
      })

      expect(textField.value).toEqual(expectedValue.toString())
    })
  })
})