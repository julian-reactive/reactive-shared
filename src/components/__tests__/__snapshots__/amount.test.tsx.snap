// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amount should match snapshot 1`] = `
<div
  class="MuiBox-root css-j9bjgs"
>
  <button
    aria-label="remove 1 unit amount"
    class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary css-b7losd-MuiButtonBase-root-MuiFab-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
      data-testid="RemoveIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13H5v-2h14v2z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <div
    class="MuiFormControl-root MuiTextField-root css-3r70aw-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      amount
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled=""
        id=":r0:"
        type="text"
        value="1"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            amount
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    aria-label="add 1 unit amount"
    class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-primary MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-primary css-1dm1o0z-MuiButtonBase-root-MuiFab-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
      data-testid="AddIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
